docs/proposals/20240309-cloneset-support-progressDeadlineSeconds.md

```diff
@@ -0,0 +1,94 @@
+---
+title: CloneSet
+authors:
+- "@hantmac"
+  reviewers:
+- "@Fei-Guo"
+- "@furykerry"
+- "@FillZpp"
+  creation-date: 2024-03-10
+  last-updated: 2024-03-10
+  status: implementable
+---
+
+# Support progressDeadlineSeconds in CloneSet
+Table of Contents
+=================
+
+- [Support progressDeadlineSeconds in CloneSet](#Support progressDeadlineSeconds in CloneSet)
+- [Table of Contents](#table-of-contents)
+  - [Motivation](#motivation)
+  - [Proposal](#proposal)
+    - [1. add .spec.progressDeadlineSeconds field](#1add-.spec.progressDeadlineSeconds-field)
+    - [2. handle the logic](#2handle-the-logic)
+
+
+## Motivation
+
+`.spec.progressDeadlineSeconds` is an optional field in Deployment that specifies the number of seconds one wants to wait for their Deployment to progress before the system reports back that the Deployment has failed progressing.
+Once the deadline has been exceeded, the Deployment controller adds a DeploymentCondition with the following attributes to the Deployment's `.status.conditions`:
+```
+type: Progressing
+status: "False"
+reason: ProgressDeadlineExceeded
+```
+
+This is useful for users to control the progress of the deployment.
+So we should add support for `progressDeadlineSeconds` in CloneSet as well. 
+
+## Proposal
+Firstly, add the `progressDeadlineSeconds` field to the CloneSetSpec. 
+Then add the handle logic in cloneSet controller to handle the `progressDeadlineSeconds` field.
+
+### 1. add .spec.progressDeadlineSeconds field
+```yaml
+apiVersion: apps.kruise.io/v1alpha1
+kind: CloneSet
+metadata:
+  name: cloneset-example
+spec:
+    replicas: 3
+    progressDeadlineSeconds: 600
+    selector:
+        matchLabels:
+        app: nginx
+    template:
+        metadata:
+        labels:
+            app: nginx
+        spec:
+        containers:
+        - name: nginx
+            image: nginx:1.14.2
+ ```
+
+### 2. handle the logic
```

Comment:
cloneset support partition, so the definition of progressDeadlineSeconds may be different from deployment. If the cloneset is paused because of partition shall we still count the paused time in progress deadline ? 
1. One possible redefinition of progressDeadlineSeconds is the time to reach completion OR "the paused state". 
2. alternatively we only support progressDeadlineSeconds if partition is not set. 
I think we need more feedback from the community

Reply:
Yes, we should not count the time when cloneset is paused by the partition.  
Only supporting `progressDeadlineSeconds` if partition is not set may be easy to implement. 
More feedback is nice. Maybe we can make a vote about these two choice.